nav2_recoveries/src/back_up.cpp

```diff
@@ -76,19 +77,46 @@ Status BackUp::onCycleUpdate()
   command_x_ < 0 ? cmd_vel.linear.x = -0.025 : cmd_vel.linear.x = 0.025;
 
   geometry_msgs::msg::Pose2D pose2d;
-  pose2d.x = current_pose.position.x + cmd_vel.linear.x * (1 / cycle_frequency_);
+  pose2d.x = current_pose.position.x;
   pose2d.y = current_pose.position.y;
   pose2d.theta = tf2::getYaw(current_pose.orientation);
 
-  if (!collision_checker_->isCollisionFree(pose2d)) {
-    stopRobot();
-    RCLCPP_WARN(node_->get_logger(), "Collision Ahead - Exiting BackUp");
-    return Status::SUCCEEDED;
+  if (!isCollisionFree(distance, cmd_vel, pose2d)) {
+    return Status::FAILED;
   }
 
   vel_publisher_->publishCommand(cmd_vel);
 
   return Status::RUNNING;
 }
 
+bool BackUp::isCollisionFree(
+  const double & distance,
+  const geometry_msgs::msg::Twist & cmd_vel,
+  geometry_msgs::msg::Pose2D & pose2d)
+{
+  // Simulate ahead by simulate_ahead_time_ in cycle_frequency_ increments
+  int cycle_count = 0;
+  double sim_position_change;
+  const double diff_dist = abs(command_x_) - distance;
+  const int max_cycle_count = static_cast<int>(cycle_frequency_ * simulate_ahead_time_);
+
+  while (cycle_count < max_cycle_count) {
+    sim_position_change = cmd_vel.linear.x * (cycle_count / cycle_frequency_);
+    pose2d.x += sim_position_change;
+    cycle_count++;
+
+    if (diff_dist - abs(sim_position_change) < 0.) {
+      break;
+    }
+
+    if (!collision_checker_->isCollisionFree(pose2d)) {
+      stopRobot();
```

Comment:
Not sure if we should stop the robot when just checking for collision.

Reply:
Commit to follow